### PR TITLE
Implementing RSA blind signatures.

### PR DESCRIPTION
**Pull request by armfazh**
Implementation of draft [draft-irtf-cfrg-rsa-blind-signatures v03](https://www.ietf.org/archive/id/draft-irtf-cfrg-rsa-blind-signatures-03.html#name-rsabssa-signature-instantia)

### CODE REVIEW

File: src/blindrsa/index.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
+import { emsa_pss_encode, i2osp, os2ip, rsasp1, rsavp1 } from './util';
+
+import sjcl from './sjcl';
+
+export class RSABlindSignProducer {
+    private kBits: number;
+    private kLen: number;
+    private hash: string;
+    constructor(public publicKey: CryptoKey) {
+        if (publicKey.type !== 'public' || publicKey.algorithm.name !== 'RSA-PSS') {
```
[the remark this answers is not in the document]
Comment (armfazh) [reply src/blindrsa/index.ts:10]: good catch.

File: src/blindrsa/index.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        const n = new sjcl.bn(Buffer.from(jwkKey.n, 'base64url').toString('hex'));
+        const e = new sjcl.bn(Buffer.from(jwkKey.e, 'base64url').toString('hex'));
+
+        // 4. r = random_integer_uniform(1, n)
+        const r = this.genRandomBlind(n);
+
+        // 5. r_inv = inverse_mod(r, n)
+        // 6. If inverse_mod fails, raise an "invalid blind" error
+        //    and stop
+        const r_inv = r.inverseMod(n);
```
[the remark this answers is not in the document]
Comment (armfazh) [reply src/blindrsa/index.ts:57]: updated now

File: src/blindrsa/index.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        const encoded_msg = await emsa_pss_encode(msg, this.kBits - 1, {
+            hash: this.hash,
+            sLen: saltLength,
+        });
+
+        // 3. m = bytes_to_int(encoded_msg)
+        const m = os2ip(encoded_msg);
+        const jwkKey = await crypto.subtle.exportKey('jwk', this.publicKey);
+        if ((jwkKey.kty && jwkKey.kty != 'RSA') || !jwkKey.n || !jwkKey.e) {
+            throw new Error('key is not rsa-pss');
```
[the remark this answers is not in the document]
Comment (armfazh) [reply src/blindrsa/index.ts:46]: done

File: src/blindrsa/index.test.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        const msg = hexToUint8(v.msg);
+        const saltLength = v.salt.length / 2;
+
+        const alice = new RSABlindSignProducer(keys.publicKey);
+        const { blinded_msg, inv } = await alice.blind(msg, saltLength);
+        expect(blinded_msg).toStrictEqual(hexToUint8(v.blinded_msg));
+        expect(inv).toStrictEqual(hexToUint8(v.inv));
+
+        const bob = new RSABlindSigner(keys.privateKey);
+        const blind_sig = await bob.sign(blinded_msg);
```
[the remark this answers is not in the document]
Comment (armfazh) [reply src/blindrsa/index.test.ts:94]: updated now

File: src/blindrsa/index.test.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+                const r_inv = hexToBn(v.inv);
+                const r = r_inv.inverseMod(n);
+                return r;
+            },
+        );
+        const keys = await keysFromVector(v);
+        const msg = hexToUint8(v.msg);
+        const saltLength = v.salt.length / 2;
+
+        const alice = new RSABlindSignProducer(keys.publicKey);
```
[the remark this answers is not in the document]
Comment (armfazh) [reply src/blindrsa/index.test.ts:88]: that's true, I changed to use functions now

File: src/blindrsa/blindrsa.test.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        // Mock for randomized blind operation.
+        jest.spyOn(crypto, 'getRandomValues')
+            .mockReturnValueOnce(hexToUint8(v.salt)) // mock for random salt
+            .mockReturnValueOnce(r_bytes); // mock for random blind
+
+        const { blinded_msg, blinded_inv } = await blindrsa.blind(publicKey, msg, saltLength);
+        expect(blinded_msg).toStrictEqual(hexToUint8(v.blinded_msg));
+        expect(blinded_inv).toStrictEqual(hexToUint8(v.inv));
+
+        const blind_sig = await blindrsa.blind_sign(privateKey, blinded_msg);
```
Comment (chris-wood) [inline src/blindrsa/blindrsa.test.ts:83]: nit: is the naming convention `blind_sign` or `blindSign`?

**Review by chris-wood — APPROVED** [review]

File: src/blindrsa/blindrsa.test.ts (hunk 6 above)
Comment (armfazh) [reply src/blindrsa/blindrsa.test.ts]: unified naming convention.

**Review by chris-wood — APPROVED** [review]

File: src/blindrsa/blindrsa.ts [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    }
+    const { modulusLength } = privateKey.algorithm as RsaHashedKeyGenParams;
+    const kLen = Math.ceil(modulusLength / 8);
+
+    const jwkKey = await crypto.subtle.exportKey('jwk', privateKey);
+    if (!jwkKey.n || !jwkKey.d) {
+        throw new Error('key is not a private key');
+    }
+    const n = new sjcl.bn(Buffer.from(jwkKey.n, 'base64url').toString('hex'));
+    const d = new sjcl.bn(Buffer.from(jwkKey.d, 'base64url').toString('hex'));
```
Comment (lukevalenta) [inline src/blindrsa/blindrsa.ts:135]: Minor nit: at https://github.com/privacypass/challenge-bypass-extension/pull/328/files#diff-80aac2cdd6b81297ebba46b540a67548f0013f4c2b574e281fc716bad65b900aR32-R37 the calculation of these intermediate values is delayed until just before their first use. If we wanted to do the same here, we could move lines https://github.com/privacypass/challenge-bypass-extension/pull/328/files#diff-80aac2cdd6b81297ebba46b540a67548f0013f4c2b574e281fc716bad65b900aR130-R135 to after step 2 below for consistency.

File: src/blindrsa/util.ts [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+        return;
+    }
+    c[0]++;
+}
+
+type MGFFn = (h: HashParams, seed: Uint8Array, mLen: number) => Promise<Uint8Array>;
+
+// MGF1
+// https://www.rfc-editor.org/rfc/rfc8017#appendix-B.2.1
+async function mgf1(h: HashParams, seed: Uint8Array, mLen: number): Promise<Uint8Array> {
```
Comment (lukevalenta) [inline src/blindrsa/util.ts:83]: The RFC says to return a `mask too long` error if `maskLen > 2^32 * hLen`: https://www.rfc-editor.org/rfc/rfc8017#appendix-B.2.1. Is it safe to omit that check here?

Comment (armfazh) [reply src/blindrsa/util.ts:83]: added now

File: src/blindrsa/blindrsa.ts (hunk 7 above)
Comment (armfazh) [reply src/blindrsa/blindrsa.ts:135]: done

File: src/blindrsa/util.ts [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+    // 1. If the message representative m is not between 0 and n - 1,
+    //    output "message representative out of range" and stop.
+    if (!m.greaterEquals(new sjcl.bn(0)) || m.greaterEquals(skS.n) == 1) {
+        throw new Error('signature representative out of range');
+    }
+    // 2. The signature representative s is computed as follows.
+    //
+    //    a. If the first form (n, d) of K is used, let s = m^d mod n.
+    const s = m.powermod(skS.d, skS.n);
+    /* TODO: implement the CRT variant.
```
[the remark this answers is not in the document]
Comment (armfazh) [reply src/blindrsa/util.ts:192]: good idea, created #329

**Review by lukevalenta — APPROVED** [review]
The new changes look good!

**Pull request merged**